# Ex 4.9 - Intro to Data Visualization with Python [Part 1]

### This script contains the following points:

##### Task 03. Import your analysis libraries, as well as your new customer data set as a dataframe.
##### Task 4. Wrangle the data so that it follows consistent logic.
##### Task 5. Complete the fundamental data quality and consistency checks.
##### Task 6. Combine your customer data with the rest of your prepared Instacart data.
##### Task 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

### Task 3. Import your analysis libraries, as well as your new customer data set as a dataframe.

In [104]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [105]:
# Turn project folder path into a string.
path = '/Users/asadagha/Documents/CF Data Analytics/Data Analytics Immersion/Achievement 4/Instacart Basket Analysis'

In [106]:
# Import new customer data set from the 'Original Data' folder as a dataframe
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [107]:
# Preview the newly imported dataframe 
customers.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [108]:
# Check the dimensions of the dataframe
customers.shape

(206209, 10)

In [109]:
# View descriptive statistics 
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [110]:
# Check data types
customers.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

### Task 4. Wrangle the data so that it follows consistent logic.

In [112]:
# Rename the columns to establish uniformity
customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)
customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)
customers.rename(columns = {'Gender' : 'gender'}, inplace = True)
customers.rename(columns = {'Age' : 'age'}, inplace = True)
customers.rename(columns = {'STATE' : 'state'}, inplace = True)
customers.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [113]:
# Check columns
customers.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'n_dependants', 'marital_status', 'income'],
      dtype='object')

In [114]:
customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

The columns have been successfully renamed and are now uniform and clear

### Task 5. Complete the fundamental data quality and consistency checks.

In [117]:
# Check for missing data
customers.isnull().sum()

user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
n_dependants          0
marital_status        0
income                0
dtype: int64

11259 missing values found in the first_name column.

In [119]:
# Check for full-duplicates
customers_dups = customers[customers.duplicated()]

In [120]:
customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, marital_status, income]
Index: []

No duplicates were found

In [122]:
# Address data-types and potential mixed data-types in the dataframe
customers.dtypes

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [123]:
# Check for columns with mixed data types
mixed_type_cols = [col for col in customers.columns if customers[col].apply(type).nunique() > 1]
if mixed_type_cols:
    print("Columns with mixed data types:", mixed_type_cols)
else:
    print("No mixed data types found.")

Columns with mixed data types: ['first_name']


In [124]:
# Convert the 'first_name' column which has mixed data-type to a string
customers['first_name'] = customers['first_name'].astype('str')

In [125]:
# The 'user_id' column is listed as an integer, change it to a string
customers['user_id'] = customers['user_id'].astype('str')

In [126]:
# Recheck the data frame and for mixed data-types
customers.dtypes

user_id           object
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
n_dependants       int64
marital_status    object
income             int64
dtype: object

In [127]:
for col in customers.columns.tolist():
  weird = (customers[[col]].map(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

All columns are the appropriate data-types and no mixed data-types are present in any columns.

In [129]:
# Check for missing data
customers.isnull().sum()

user_id           0
first_name        0
last_name         0
gender            0
state             0
age               0
date_joined       0
n_dependants      0
marital_status    0
income            0
dtype: int64

In [130]:
# Perform a descriptive statistic check to ensure everything looks good
customers.describe()

age   n_dependants         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

### Task 6. Combine your customer data with the rest of your prepared Instacart data.

In [132]:
# Import the rest of the prepared Instacart data
ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_grouped.pkl'))

In [133]:
# Check the shape of the newly imported dataframe
ords_prods.shape

(32404859, 26)

In [134]:
# Check the shape of the customers dataframe
customers.shape

(206209, 10)

In [135]:
# Identify a shared column
ords_prods.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

     price_range_loc     busiest_day    busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy            Most orders   
1  Mid-range product  Regularly busy  Regularly busy         Average orders   
2  Mid-range product     Busiest day    Busiest days         Average orders   
3  Mid-range product  Regularly busy    Slowest days            Most orders   
4  Mid-range product      Least busy    Slowest days         Average orders   

  max_order      loyalty_flag average_spend spender_flag customer_frequency  \
0        32  Regular customer      6.935811  Low_spender                8.0   
1        32  Regular customer      6.935811  Low_spender                8.0   
2         5      New customer      7.930208  Low_spender                8.0   
3         3      New customer      4.972414  Low_spender                9.0   
4         3      New customer      4.972414  Low_spender                9.0   

      frequency_flag  
0  Frequent customer  
1  Frequent customer  
2  Frequent customer  
3  Frequent customer  
4  Frequent customer  

[5 rows x 26 columns]

In [136]:
customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

In [137]:
# Both dataframes share the 'user_id' column, check data-type of shared column.
ords_prods.dtypes['user_id']

dtype('int64')

In [138]:
customers.dtypes['user_id']

dtype('O')

In [139]:
# Make sure both key columns are the same data type
ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [140]:
ords_prods.dtypes['user_id']

dtype('O')

The key columns now have matching data-types

In [142]:
# Drop the existing '_merge' column from previous exercises
ords_prods = ords_prods.drop(columns=['_merge'])

In [143]:
# Drop the existing 'merge_dataframes' column from previous exercises
ords_prods = ords_prods.drop(columns=['merged_dataframes'])

In [144]:
# Merge the two dataframes using the column 'user_id' as the shared column with the default join (inner)
customers_ords_prods = ords_prods.merge(customers, on = 'user_id', indicator = True)

In [145]:
# Check shape of newly merged dataframe
customers_ords_prods.shape

(32404859, 34)

In [146]:
customers_ords_prods.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998     138            28                  6                 11  ...   
1   1977647     138            30                  6                 17  ...   
2    389851     709             2                  0                 21  ...   
3    652770     764             1                  3                 13  ...   
4   1813452     764             3                  4                 17  ...   

   first_name  last_name  gender      state age date_joined n_dependants  \
0     Charles        Cox    Male  Minnesota  81    8/1/2019            1   
1     Charles        Cox    Male  Minnesota  81    8/1/2019            1   
2     Deborah      Glass  Female    Vermont  66   6/16/2018            2   
3     Heather      Myers  Female  Wisconsin  40    2/9/2020            3   
4     Heather      Myers  Female  Wisconsin  40    2/9/2020            3   

  marital_status  income _merge  
0        married   49620   both  
1        married   49620   both  
2        married  158302   both  
3        married   31308   both  
4        married   31308   both  

[5 rows x 34 columns]

In [147]:
# Check inner join value counts
customers_ords_prods['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [148]:
# Check data-types
customers_ords_prods.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                     int64
user_id                     object
order_number                 int64
order_day_of_week            int64
order_hour_of_day            int64
days_since_prior_order     float64
first_order                   bool
add_to_cart_order            int64
reordered                    int64
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_spend              float64
spender_flag                object
customer_frequency         float64
frequency_flag              object
first_name                  object
last_name                   object
gender                      object
state                       object
age                 

In [187]:
# Change the 'order_id' column to a string data type
customers_ords_prods['order_id'] = customers_ords_prods['order_id'].astype('str')

In [191]:
customers_ords_prods.dtypes

product_id                   int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
order_id                    object
user_id                     object
order_number                 int64
order_day_of_week            int64
order_hour_of_day            int64
days_since_prior_order     float64
first_order                   bool
add_to_cart_order            int64
reordered                    int64
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
average_spend              float64
spender_flag                object
customer_frequency         float64
frequency_flag              object
first_name                  object
last_name                   object
gender                      object
state                       object
age                 

### Task 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [197]:
customers_ords_prods.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customers_orders_products_merged.pkl'))